File: mobile/lib/domain/ride/ride_telemetry.dart
import '../ride.dart';
import '../routing/route_progress.dart';

/// Post-Ride-Telemetrie — Spiegel [src/lib/ride/rideTelemetry.ts].
/// GPS-Höhe rauschig: Fenster-Neigung, Lücken bleiben Lücken.

enum GradeBand { steepUp, up, roll, down, steepDown, gap }

class RideSample {
  const RideSample({
    required this.distKm,
    required this.lat,
    required this.lng,
    this.elevM,
    this.gradePct,
    required this.band,
    this.speedKmh,
    this.hr,
    this.cad,
    this.power,
    this.lean,
    this.gPeak,
    this.impact = false,
  });

  final double distKm;
  final double lat;
  final double lng;
  final double? elevM;
  final double? gradePct;
  final GradeBand band;
  final double? speedKmh;
  final int? hr;
  final int? cad;
  final int? power;
  final double? lean;
  final double? gPeak;
  final bool impact;
}

class RideChannels {
  const RideChannels({
    required this.elev,
    required this.speed,
    required this.hr,
    required this.cad,
    required this.power,
    required this.lean,
    required this.g,
    required this.impact,
  });

  final bool elev;
  final bool speed;
  final bool hr;
  final bool cad;
  final bool power;
  final bool lean;
  final bool g;
  final bool impact;
}

class GradeLine {
  const GradeLine({
    required this.points,
    required this.colorHex,
    required this.band,
  });

  final List<({double lat, double lng})> points;
  final String colorHex;
  final GradeBand band;
}

class RideTelemetry {
  const RideTelemetry({
    required this.samples,
    required this.chart,
    required this.totalDistKm,
    required this.climbM,
    required this.descentM,
    required this.gapKm,
    this.maxGradePct,
    this.minGradePct,
    this.maxSpeedKmh,
    this.avgSpeedKmh,
    this.avgHr,
    this.maxHr,
    this.avgCad,
    this.avgPower,
    required this.impactCount,
    this.maxLean,
    this.maxG,
    required this.channels,
  });

  final List<RideSample> samples;
  final List<RideSample> chart;
  final double totalDistKm;
  final int climbM;
  final int descentM;
  final double gapKm;
  final double? maxGradePct;
  final double? minGradePct;
  final double? maxSpeedKmh;
  final double? avgSpeedKmh;
  final int? avgHr;
  final int? maxHr;
  final int? avgCad;
  final int? avgPower;
  final int impactCount;
  final double? maxLean;
  final double? maxG;
  final RideChannels channels;

  bool get hasElev => channels.elev;
}

const _gradeWindowM = 35.0;
const _gradeMinM = 18.0;
const _smoothRadiusM = 40.0;
const _climbStepM = 1.2;
const _maxAbsGrade = 45.0;
const _maxSpeedKmh = 85.0;
const _chartMax = 240;
const _mapMax = 90;

const Map<GradeBand, String> kGradeColors = {
  GradeBand.steepUp: '#C2410C',
  GradeBand.up: '#FF6A00',
  GradeBand.roll: '#7A8B73',
  GradeBand.down: '#5B8C9A',
  GradeBand.steepDown: '#3D6B8A',
  GradeBand.gap: '#6B7280',
};

GradeBand gradeBand(double? gradePct) {
  if (gradePct == null || gradePct.isNaN) return GradeBand.gap;
  if (gradePct > 8) return GradeBand.steepUp;
  if (gradePct > 3) return GradeBand.up;
  if (gradePct >= -3) return GradeBand.roll;
  if (gradePct >= -8) return GradeBand.down;
  return GradeBand.steepDown;
}

RideTelemetry emptyRideTelemetry() => const RideTelemetry(
      samples: [],
      chart: [],
      totalDistKm: 0,
      climbM: 0,
      descentM: 0,
      gapKm: 0,
      impactCount: 0,
      channels: RideChannels(
        elev: false,
        speed: false,
        hr: false,
        cad: false,
        power: false,
        lean: false,
        g: false,
        impact: false,
      ),
    );

RideTelemetry buildRideTelemetry(List<Map<String, dynamic>> track) {
  if (track.length < 2) return emptyRideTelemetry();

  final raw = <_Raw>[];
  for (final p in track) {
    final lat = (p['lat'] as num?)?.toDouble() ??
        (p['latitude'] as num?)?.toDouble();
    final lng = (p['lng'] as num?)?.toDouble() ??
        (p['lon'] as num?)?.toDouble() ??
        (p['longitude'] as num?)?.toDouble();
    if (lat == null || lng == null) continue;
    if (lat.abs() < 1e-6 && lng.abs() < 1e-6) continue;
    final elev = _elev(p);
    raw.add(
      _Raw(
        lat: lat,
        lng: lng,
        elev: elev,
        timeMs: _timeMs(p),
        hr: liveHrFromTrackPoint(p),
        cad: liveCadFromTrackPoint(p),
        power: livePowerFromTrackPoint(p),
        lean: liveLeanFromTrackPoint(p),
        gPeak: liveGFromTrackPoint(p),
        impact: liveImpactFromTrackPoint(p),
        spd: liveSpeedFromTrackPoint(p),
      ),
    );
  }
  if (raw.length < 2) return emptyRideTelemetry();

  final distM = List<double>.filled(raw.length, 0);
  for (var i = 1; i < raw.length; i++) {
    distM[i] = distM[i - 1] +
        haversineM(raw[i - 1].lat, raw[i - 1].lng, raw[i].lat, raw[i].lng);
  }

  final smoothed = List<double?>.filled(raw.length, null);
  for (var i = 0; i < raw.length; i++) {
    final e = raw[i].elev;
    if (e == null) continue;
    final lo = distM[i] - _smoothRadiusM;
    final hi = distM[i] + _smoothRadiusM;
    var sum = 0.0;
    var n = 0;
    for (var j = 0; j < raw.length; j++) {
      final ej = raw[j].elev;
      if (ej == null) continue;
      if (distM[j] < lo) continue;
      if (distM[j] > hi) break;
      sum += ej;
      n += 1;
    }
    smoothed[i] = n == 0 ? e : sum / n;
  }

  var climbM = 0.0;
  var descentM = 0.0;
  var gapKm = 0.0;
  var prevElev = smoothed[0];
  var prevDist = 0.0;
  for (var i = 1; i < raw.length; i++) {
    final e = smoothed[i];
    final dKm = (distM[i] - prevDist) / 1000;
    if (e == null) {
      gapKm += dKm;
      prevDist = distM[i];
      continue;
    }
    if (prevElev != null) {
      final d = e - prevElev;
      if (d > _climbStepM) {
        climbM += d;
      } else if (d < -_climbStepM) {
        descentM += -d;
      }
    }
    prevElev = e;
    prevDist = distM[i];
  }

  final samples = <RideSample>[];
  for (var i = 0; i < raw.length; i++) {
    double? gradePct;
    if (smoothed[i] != null && i > 0) {
      final j = _lookBack(distM, i, _gradeWindowM);
      final span = distM[i] - distM[j];
      final a = smoothed[j];
      final b = smoothed[i];
      if (a != null && b != null && span >= _gradeMinM) {
        final g = ((b - a) / span) * 100;
        if (g.isFinite) {
          gradePct = ((g.clamp(-_maxAbsGrade, _maxAbsGrade)) * 10).round() / 10;
        }
      }
    }

    var speed = raw[i].spd;
    final t0 = raw[i].timeMs;
    final t1 = i > 0 ? raw[i - 1].timeMs : null;
    if (speed == null && t0 != null && t1 != null) {
      final dt = (t0 - t1) / 1000.0;
      final hop = distM[i] - distM[i - 1];
      final maxDt = hop >= 15 ? 900 : 30;
      if (dt >= 0.4 && dt <= maxDt && hop >= 0) {
        final v = (hop / dt) * 3.6;
        if (v >= 0.4 && v <= _maxSpeedKmh) {
          speed = (v * 10).round() / 10;
        }
      }
    }

    samples.add(
      RideSample(
        distKm: ((distM[i] / 1000) * 1000).round() / 1000,
        lat: raw[i].lat,
        lng: raw[i].lng,
        elevM: smoothed[i] == null
            ? null
            : (smoothed[i]! * 10).round() / 10,
        gradePct: gradePct,
        band: gradeBand(gradePct),
        speedKmh: speed,
        hr: raw[i].hr,
        cad: raw[i].cad,
        power: raw[i].power,
        lean: raw[i].lean,
        gPeak: raw[i].gPeak,
        impact: raw[i].impact,
      ),
    );
  }

  final elevs = [for (final s in samples) if (s.elevM != null) s.elevM!];
  final grades = [for (final s in samples) if (s.gradePct != null) s.gradePct!];
  final speeds = [for (final s in samples) if (s.speedKmh != null) s.speedKmh!];
  final hrs = [for (final s in samples) if (s.hr != null) s.hr!];
  final cads = [for (final s in samples) if (s.cad != null) s.cad!];
  final powers = [for (final s in samples) if (s.power != null) s.power!];
  final leans = [for (final s in samples) if (s.lean != null) s.lean!];
  final gs = [for (final s in samples) if (s.gPeak != null) s.gPeak!];
  final impacts = samples.where((s) => s.impact).length;

  final channels = RideChannels(
    elev: elevs.length >= 2,
    speed: speeds.length >= 2,
    hr: hrs.length >= 2,
    cad: cads.length >= 2,
    power: powers.length >= 2,
    lean: leans.length >= 2,
    g: gs.length >= 2,
    impact: impacts > 0,
  );

  return RideTelemetry(
    samples: samples,
    chart: _downsample(samples, _chartMax, (s) => s.impact),
    totalDistKm: ((distM.last / 1000) * 100).round() / 100,
    climbM: climbM.round(),
    descentM: descentM.round(),
    gapKm: (gapKm * 100).round() / 100,
    maxGradePct: grades.isEmpty ? null : grades.reduce((a, b) => a > b ? a : b),
    minGradePct: grades.isEmpty ? null : grades.reduce((a, b) => a < b ? a : b),
    maxSpeedKmh: speeds.isEmpty ? null : speeds.reduce((a, b) => a > b ? a : b),
    avgSpeedKmh: _mean(speeds),
    avgHr: hrs.isEmpty ? null : (_sumInt(hrs) / hrs.length).round(),
    maxHr: hrs.isEmpty ? null : hrs.reduce((a, b) => a > b ? a : b),
    avgCad: cads.isEmpty ? null : (_sumInt(cads) / cads.length).round(),
    avgPower: powers.isEmpty ? null : (_sumInt(powers) / powers.length).round(),
    impactCount: impacts,
    maxLean: leans.isEmpty
        ? null
        : leans.map((e) => e.abs()).reduce((a, b) => a > b ? a : b),
    maxG: gs.isEmpty ? null : gs.reduce((a, b) => a > b ? a : b),
    channels: channels,
  );
}

List<GradeLine> gradeMapLayers(RideTelemetry telemetry) {
  final pts = _downsample(telemetry.samples, _mapMax, (_) => false);
  if (pts.length < 2) return const [];
  final out = <GradeLine>[];
  var start = 0;
  var band = pts.first.band;
  void flush(int end) {
    final slice = pts.sublist(start, end + 1);
    if (slice.length < 2) return;
    out.add(
      GradeLine(
        points: [
          for (final p in slice) (lat: p.lat, lng: p.lng),
        ],
        colorHex: kGradeColors[band]!,
        band: band,
      ),
    );
  }

  for (var i = 1; i < pts.length; i++) {
    if (pts[i].band != band) {
      flush(i);
      start = i;
      band = pts[i].band;
    }
  }
  flush(pts.length - 1);
  return out;
}

class _Raw {
  const _Raw({
    required this.lat,
    required this.lng,
    required this.elev,
    required this.timeMs,
    required this.hr,
    required this.cad,
    required this.power,
    required this.lean,
    required this.gPeak,
    required this.impact,
    required this.spd,
  });

  final double lat;
  final double lng;
  final double? elev;
  final int? timeMs;
  final int? hr;
  final int? cad;
  final int? power;
  final double? lean;
  final double? gPeak;
  final bool impact;
  final double? spd;
}

double? _elev(Map<String, dynamic> p) {
  final v = p['elev'] ?? p['ele'] ?? p['altitude'];
  if (v is! num) return null;
  final n = v.toDouble();
  if (n < -50 || n > 8900) return null;
  return n;
}

int? _timeMs(Map<String, dynamic> p) {
  final v = p['time'] ?? p['timeMs'];
  if (v is! num) return null;
  final n = v.toDouble();
  if (n >= 1e12) return n.round();
  if (n >= 1e9) return (n * 1000).round();
  return n.round();
}

int _lookBack(List<double> distM, int i, double windowM) {
  var j = i;
  while (j > 0 && distM[i] - distM[j] < windowM) {
    j -= 1;
  }
  return j;
}

List<T> _downsample<T>(
  List<T> items,
  int max,
  bool Function(T) keep,
) {
  if (items.length <= max) return items;
  final taken = <int>{};
  final step = (items.length - 1) / (max - 1);
  for (var i = 0; i < max; i++) {
    taken.add((i * step).round());
  }
  for (var i = 0; i < items.length; i++) {
    if (keep(items[i])) taken.add(i);
  }
  final idxs = taken.toList()..sort();
  return [for (final i in idxs) items[i]];
}

double? _mean(List<double> values) {
  if (values.isEmpty) return null;
  return (values.reduce((a, b) => a + b) / values.length * 10).round() / 10;
}

/// Persistierte Höhenmeter: Telemetrie-Anstieg, sonst gespeicherter Wert.
int honestClimbM(List<Map<String, dynamic>> track, [num storedM = 0]) {
  final tel = buildRideTelemetry(track);
  if (tel.hasElev) return tel.climbM;
  final s = storedM.round();
  return s > 0 ? s : 0;
}

RideSample? nearestSample(RideTelemetry telemetry, double distKm) {
  final pts = telemetry.chart.isNotEmpty ? telemetry.chart : telemetry.samples;
  if (pts.isEmpty) return null;
  var best = pts.first;
  var bestD = (best.distKm - distKm).abs();
  for (final p in pts.skip(1)) {
    final d = (p.distKm - distKm).abs();
    if (d < bestD) {
      best = p;
      bestD = d;
    }
  }
  return best;
}

/// Kurze Zeile für Peek — nur was der Track wirklich trägt.
String? terrainCaption(RideTelemetry telemetry, [String hm = 'hm']) {
  if (!telemetry.hasElev) return null;
  final parts = <String>['${telemetry.climbM} $hm'];
  if (telemetry.descentM > 0) parts.add('${telemetry.descentM} $hm ↓');
  if (telemetry.maxGradePct != null) {
    final g = telemetry.maxGradePct!;
    parts.add('${g > 0 ? '+' : ''}${g.toStringAsFixed(0)} %');
  }
  return parts.join(' · ');
}

int _sumInt(List<int> values) => values.fold(0, (a, b) => a + b);

File: src/lib/ride/rideTelemetry.ts
/**
 * Post-Ride-Telemetrie aus dem aufgezeichneten Track.
 *
 * GPS-Höhe ist rauschig — deshalb Fenster-Neigung, leichte Glättung,
 * Lücken bleiben Lücken. Kein DEM, kein Interpolieren als Messung.
 * Sensor-Kanäle (Puls, Cadenz, Leistung, Lean, g, Impact) nur wenn
 * sie wirklich auf Punkten stehen.
 */

export type GradeBand =
  | "steep_up"
  | "up"
  | "roll"
  | "down"
  | "steep_down"
  | "gap";

export const GRADE_COLORS: Record<GradeBand, string> = {
  steep_up: "#C2410C",
  up: "#FF6A00",
  roll: "#7A8B73",
  down: "#5B8C9A",
  steep_down: "#3D6B8A",
  gap: "#6B7280",
};

export type RideTrackInput = {
  lat?: number;
  latitude?: number;
  lng?: number;
  lon?: number;
  longitude?: number;
  elev?: number | null;
  ele?: number | null;
  altitude?: number | null;
  time?: number;
  timeMs?: number;
  hr?: number;
  heartRateBpm?: number;
  cad?: number;
  cadenceRpm?: number;
  power?: number;
  powerW?: number;
  lean?: number;
  leanDeg?: number;
  g?: number;
  gPeak?: number;
  impact?: number | boolean;
  spd?: number;
  speedKmh?: number;
};

export type RideSample = {
  distKm: number;
  lat: number;
  lng: number;
  elevM: number | null;
  gradePct: number | null;
  band: GradeBand;
  speedKmh: number | null;
  hr: number | null;
  cad: number | null;
  power: number | null;
  lean: number | null;
  gPeak: number | null;
  impact: boolean;
};

export type RideChannels = {
  elev: boolean;
  speed: boolean;
  hr: boolean;
  cad: boolean;
  power: boolean;
  lean: boolean;
  g: boolean;
  impact: boolean;
};

export type GradeLine = {
  id: string;
  coordinates: [number, number][];
  color: string;
  band: GradeBand;
};

export type RideTelemetry = {
  samples: RideSample[];
  chart: RideSample[];
  totalDistKm: number;
  climbM: number;
  descentM: number;
  gapKm: number;
  maxGradePct: number | null;
  minGradePct: number | null;
  maxSpeedKmh: number | null;
  avgSpeedKmh: number | null;
  avgHr: number | null;
  maxHr: number | null;
  avgCad: number | null;
  avgPower: number | null;
  impactCount: number;
  maxLean: number | null;
  maxG: number | null;
  channels: RideChannels;
  elevSource: "gps" | "none";
};

const R_EARTH_M = 6_371_000;
const GRADE_WINDOW_M = 35;
const GRADE_MIN_M = 18;
const SMOOTH_RADIUS_M = 40;
const CLIMB_STEP_M = 1.2;
const MAX_ABS_GRADE = 45;
const MAX_SPEED_KMH = 85;
const CHART_MAX = 240;
const MAP_MAX = 90;

export function haversineM(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number
): number {
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLng / 2) ** 2;
  return 2 * R_EARTH_M * Math.asin(Math.min(1, Math.sqrt(a)));
}

function toRad(d: number): number {
  return (d * Math.PI) / 180;
}

function finiteNum(v: unknown): number | null {
  if (typeof v !== "number" || !Number.isFinite(v)) return null;
  return v;
}

function readLat(p: RideTrackInput): number | null {
  return finiteNum(p.lat) ?? finiteNum(p.latitude);
}

function readLng(p: RideTrackInput): number | null {
  return finiteNum(p.lng) ?? finiteNum(p.lon) ?? finiteNum(p.longitude);
}

function readElev(p: RideTrackInput): number | null {
  const raw =
    finiteNum(p.elev) ?? finiteNum(p.ele) ?? finiteNum(p.altitude);
  if (raw == null || raw < -50 || raw > 8900) return null;
  return raw;
}

function readHr(p: RideTrackInput): number | null {
  const v = finiteNum(p.hr) ?? finiteNum(p.heartRateBpm);
  if (v == null || v < 1 || v > 239) return null;
  return Math.round(v);
}

function readCad(p: RideTrackInput): number | null {
  const v = finiteNum(p.cad) ?? finiteNum(p.cadenceRpm);
  if (v == null || v < 1 || v > 254) return null;
  return Math.round(v);
}

function readPower(p: RideTrackInput): number | null {
  const v = finiteNum(p.power) ?? finiteNum(p.powerW);
  if (v == null || v < 1 || v > 2500) return null;
  return Math.round(v);
}

function readLean(p: RideTrackInput): number | null {
  const v = finiteNum(p.lean) ?? finiteNum(p.leanDeg);
  if (v == null || Math.abs(v) > 80) return null;
  return Math.round(v * 10) / 10;
}

function readG(p: RideTrackInput): number | null {
  const v = finiteNum(p.g) ?? finiteNum(p.gPeak);
  if (v == null || v <= 0 || v > 20) return null;
  return Math.round(v * 100) / 100;
}

function readImpact(p: RideTrackInput): boolean {
  if (p.impact === true) return true;
  const v = finiteNum(p.impact as number);
  return v != null && v > 0;
}

function readStampedSpeed(p: RideTrackInput): number | null {
  const v = finiteNum(p.spd) ?? finiteNum(p.speedKmh);
  if (v == null || v < 0.4 || v > MAX_SPEED_KMH) return null;
  return Math.round(v * 10) / 10;
}

function readTimeMs(p: RideTrackInput): number | null {
  const raw = finiteNum(p.time) ?? finiteNum(p.timeMs);
  if (raw == null) return null;
  if (raw >= 1e12) return raw;
  if (raw >= 1e9) return raw * 1000;
  return raw;
}

export function gradeBand(gradePct: number | null): GradeBand {
  if (gradePct == null || !Number.isFinite(gradePct)) return "gap";
  if (gradePct > 8) return "steep_up";
  if (gradePct > 3) return "up";
  if (gradePct >= -3) return "roll";
  if (gradePct >= -8) return "down";
  return "steep_down";
}

function mean(values: number[]): number | null {
  if (values.length === 0) return null;
  return values.reduce((a, b) => a + b, 0) / values.length;
}

function maxOf(values: number[]): number | null {
  if (values.length === 0) return null;
  return Math.max(...values);
}

function minOf(values: number[]): number | null {
  if (values.length === 0) return null;
  return Math.min(...values);
}

function lookBackIndex(distM: number[], i: number, windowM: number): number {
  let j = i;
  while (j > 0 && distM[i] - distM[j] < windowM) j -= 1;
  return j;
}

function downsample<T>(
  items: T[],
  max: number,
  keep?: (item: T) => boolean
): T[] {
  if (items.length <= max) return items;
  const taken = new Set<number>();
  const step = (items.length - 1) / (max - 1);
  for (let i = 0; i < max; i++) {
    taken.add(Math.round(i * step));
  }
  if (keep) {
    for (let i = 0; i < items.length; i++) {
      if (keep(items[i])) taken.add(i);
    }
  }
  return [...taken].sort((a, b) => a - b).map((i) => items[i]);
}

function emptyTelemetry(): RideTelemetry {
  return {
    samples: [],
    chart: [],
    totalDistKm: 0,
    climbM: 0,
    descentM: 0,
    gapKm: 0,
    maxGradePct: null,
    minGradePct: null,
    maxSpeedKmh: null,
    avgSpeedKmh: null,
    avgHr: null,
    maxHr: null,
    avgCad: null,
    avgPower: null,
    impactCount: 0,
    maxLean: null,
    maxG: null,
    channels: {
      elev: false,
      speed: false,
      hr: false,
      cad: false,
      power: false,
      lean: false,
      g: false,
      impact: false,
    },
    elevSource: "none",
  };
}

export function buildRideTelemetry(
  track: RideTrackInput[] | undefined | null
): RideTelemetry {
  if (!track || track.length < 2) return emptyTelemetry();

  const raw: {
    lat: number;
    lng: number;
    elev: number | null;
    timeMs: number | null;
    hr: number | null;
    cad: number | null;
    power: number | null;
    lean: number | null;
    gPeak: number | null;
    impact: boolean;
    spd: number | null;
  }[] = [];

  for (const p of track) {
    const lat = readLat(p);
    const lng = readLng(p);
    if (lat == null || lng == null) continue;
    if (Math.abs(lat) < 1e-6 && Math.abs(lng) < 1e-6) continue;
    raw.push({
      lat,
      lng,
      elev: readElev(p),
      timeMs: readTimeMs(p),
      hr: readHr(p),
      cad: readCad(p),
      power: readPower(p),
      lean: readLean(p),
      gPeak: readG(p),
      impact: readImpact(p),
      spd: readStampedSpeed(p),
    });
  }

  if (raw.length < 2) return emptyTelemetry();

  const distM = new Array<number>(raw.length);
  distM[0] = 0;
  for (let i = 1; i < raw.length; i++) {
    distM[i] =
      distM[i - 1] +
      haversineM(raw[i - 1].lat, raw[i - 1].lng, raw[i].lat, raw[i].lng);
  }

  const smoothed = raw.map((p, i) => {
    if (p.elev == null) return null;
    const lo = distM[i] - SMOOTH_RADIUS_M;
    const hi = distM[i] + SMOOTH_RADIUS_M;
    const win: number[] = [];
    for (let j = 0; j < raw.length; j++) {
      const e = raw[j].elev;
      if (e == null) continue;
      if (distM[j] < lo) continue;
      if (distM[j] > hi) break;
      win.push(e);
    }
    if (win.length === 0) return p.elev;
    return win.reduce((a, b) => a + b, 0) / win.length;
  });

  let climbM = 0;
  let descentM = 0;
  let gapKm = 0;
  let prevElev: number | null = smoothed[0];
  let prevDist = 0;
  for (let i = 1; i < raw.length; i++) {
    const e = smoothed[i];
    const dKm = (distM[i] - prevDist) / 1000;
    if (e == null) {
      gapKm += dKm;
      prevDist = distM[i];
      continue;
    }
    if (prevElev != null) {
      const d = e - prevElev;
      if (d > CLIMB_STEP_M) climbM += d;
      else if (d < -CLIMB_STEP_M) descentM += -d;
    }
    prevElev = e;
    prevDist = distM[i];
  }

  const samples: RideSample[] = raw.map((p, i) => {
    let gradePct: number | null = null;
    if (smoothed[i] != null && i > 0) {
      const j = lookBackIndex(distM, i, GRADE_WINDOW_M);
      const span = distM[i] - distM[j];
      const a = smoothed[j];
      const b = smoothed[i];
      if (a != null && b != null && span >= GRADE_MIN_M) {
        const g = ((b - a) / span) * 100;
        if (Number.isFinite(g)) {
          gradePct = Math.round(Math.max(-MAX_ABS_GRADE, Math.min(MAX_ABS_GRADE, g)) * 10) / 10;
        }
      }
    }

    let speedKmh = p.spd;
    if (speedKmh == null && i > 0 && p.timeMs != null && raw[i - 1].timeMs != null) {
      const dt = (p.timeMs - raw[i - 1].timeMs!) / 1000;
      const hop = distM[i] - distM[i - 1];
      const maxDt = hop >= 15 ? 900 : 30;
      if (dt >= 0.4 && dt <= maxDt && hop >= 0) {
        const v = (hop / dt) * 3.6;
        if (v >= 0.4 && v <= MAX_SPEED_KMH) {
          speedKmh = Math.round(v * 10) / 10;
        }
      }
    }

    return {
      distKm: Math.round((distM[i] / 1000) * 1000) / 1000,
      lat: p.lat,
      lng: p.lng,
      elevM: smoothed[i] != null ? Math.round(smoothed[i]! * 10) / 10 : null,
      gradePct,
      band: gradeBand(gradePct),
      speedKmh,
      hr: p.hr,
      cad: p.cad,
      power: p.power,
      lean: p.lean,
      gPeak: p.gPeak,
      impact: p.impact,
    };
  });

  const elevs = samples.map((s) => s.elevM).filter((e): e is number => e != null);
  const grades = samples
    .map((s) => s.gradePct)
    .filter((g): g is number => g != null);
  const speeds = samples
    .map((s) => s.speedKmh)
    .filter((v): v is number => v != null);
  const hrs = samples.map((s) => s.hr).filter((v): v is number => v != null);
  const cads = samples.map((s) => s.cad).filter((v): v is number => v != null);
  const powers = samples
    .map((s) => s.power)
    .filter((v): v is number => v != null);
  const leans = samples.map((s) => s.lean).filter((v): v is number => v != null);
  const gs = samples.map((s) => s.gPeak).filter((v): v is number => v != null);
  const impactCount = samples.filter((s) => s.impact).length;

  const channels: RideChannels = {
    elev: elevs.length >= 2,
    speed: speeds.length >= 2,
    hr: hrs.length >= 2,
    cad: cads.length >= 2,
    power: powers.length >= 2,
    lean: leans.length >= 2,
    g: gs.length >= 2,
    impact: impactCount > 0,
  };

  const totalDistKm = Math.round((distM[distM.length - 1] / 1000) * 100) / 100;
  const chart = downsample(samples, CHART_MAX, (s) => s.impact);

  return {
    samples,
    chart,
    totalDistKm,
    climbM: Math.round(climbM),
    descentM: Math.round(descentM),
    gapKm: Math.round(gapKm * 100) / 100,
    maxGradePct: maxOf(grades),
    minGradePct: minOf(grades),
    maxSpeedKmh: maxOf(speeds),
    avgSpeedKmh: mean(speeds) != null ? Math.round(mean(speeds)! * 10) / 10 : null,
    avgHr: mean(hrs) != null ? Math.round(mean(hrs)!) : null,
    maxHr: maxOf(hrs),
    avgCad: mean(cads) != null ? Math.round(mean(cads)!) : null,
    avgPower: mean(powers) != null ? Math.round(mean(powers)!) : null,
    impactCount,
    maxLean: maxOf(leans.map(Math.abs)),
    maxG: maxOf(gs),
    channels,
    elevSource: channels.elev ? "gps" : "none",
  };
}

export function gradeMapLayers(telemetry: RideTelemetry): GradeLine[] {
  const pts = downsample(telemetry.samples, MAP_MAX);
  if (pts.length < 2) return [];
  const lines: GradeLine[] = [];
  let start = 0;
  let band = pts[0].band;
  let n = 0;
  const flush = (end: number) => {
    const slice = pts.slice(start, end + 1);
    if (slice.length < 2) return;
    lines.push({
      id: `grade-${n++}`,
      coordinates: slice.map((p) => [p.lng, p.lat]),
      color: GRADE_COLORS[band],
      band,
    });
  };
  for (let i = 1; i < pts.length; i++) {
    if (pts[i].band !== band) {
      flush(i);
      start = i;
      band = pts[i].band;
    }
  }
  flush(pts.length - 1);
  return lines;
}

/** Persistierte Höhenmeter: Telemetrie-Anstieg, sonst gespeicherter Wert. */
export function honestClimbM(
  track: RideTrackInput[] | undefined | null,
  storedM = 0
): number {
  const tel = buildRideTelemetry(track);
  if (tel.channels.elev) return tel.climbM;
  return typeof storedM === "number" && Number.isFinite(storedM) && storedM > 0
    ? Math.round(storedM)
    : 0;
}

export function nearestSample(
  telemetry: RideTelemetry,
  distKm: number
): RideSample | null {
  const pts = telemetry.chart.length > 0 ? telemetry.chart : telemetry.samples;
  if (pts.length === 0) return null;
  let best = pts[0];
  let bestD = Math.abs(best.distKm - distKm);
  for (let i = 1; i < pts.length; i++) {
    const d = Math.abs(pts[i].distKm - distKm);
    if (d < bestD) {
      best = pts[i];
      bestD = d;
    }
  }
  return best;
}
